contracts/chain-adapters/l2/Ovm_WithdrawalAdapter.sol

```diff
@@ -0,0 +1,138 @@
+
+import { WithdrawalAdapter, ITokenMessenger } from "./WithdrawalAdapter.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { WETH9Interface } from "../../external/interfaces/WETH9Interface.sol";
+import { Lib_PredeployAddresses } from "@eth-optimism/contracts/libraries/constants/Lib_PredeployAddresses.sol";
+import { IL2ERC20Bridge } from "../../Ovm_SpokePool.sol";
+
+/**
+ * @notice OVM specific bridge adapter. Implements logic to bridge tokens back to mainnet.
+ * @custom:security-contact bugs@across.to
+ */
+
+interface IOvm_SpokePool {
+    // @dev Returns the address of the token bridge for the input l2 token.
+    function tokenBridges(address token) external view returns (address);
+
+    // @dev Returns the address of the l1 token set in the spoke pool for the input l2 token.
+    function remoteL1Tokens(address token) external view returns (address);
+
+    // @dev Returns the address for the representation of ETH on the l2.
+    function l2Eth() external view returns (address);
+
+    // @dev Returns the address of the wrapped native token for the L2.
+    function wrappedNativeToken() external view returns (WETH9Interface);
+
+    // @dev Returns the amount of gas the contract allocates for a token withdrawal.
+    function l1Gas() external view returns (uint32);
+}
+
+/**
+ * @title Adapter for interacting with bridges from an OpStack L2 to Ethereum mainnet.
+ * @notice This contract is used to share L2-L1 bridging logic with other L2 Across contracts.
+ */
+contract Ovm_WithdrawalAdapter is WithdrawalAdapter {
+    using SafeERC20 for IERC20;
+
+    // Address for the wrapped native token on this chain. For Ovm standard bridges, we need to unwrap
+    // this token before initiating the withdrawal. Normally, it is 0x42..006, but there are instances
+    // where this address is different.
+    WETH9Interface public immutable wrappedNativeToken;
+    // Address which represents the native token on L2. For OpStack chains, this is generally 0xDeadDeAdde...aDDeAD0000.
+    address public immutable l2Eth;
+    // Stores required gas to send tokens back to L1.
+    uint32 public immutable l1Gas;
+    // Address of the corresponding spoke pool on L2. This is to piggyback off of the spoke pool's supported
+    // token routes/defined token bridges.
+    IOvm_SpokePool public immutable spokePool;
+
+    /*
+     * @notice constructs the withdrawal adapter.
+     * @param _l2Usdc address of native USDC on the L2.
+     * @param _cctpTokenMessenger address of the CCTP token messenger contract on L2.
+     * @param _l2Gateway address of the Optimism ERC20 l2 standard bridge contract.
+     */
+    constructor(
+        IERC20 _l2Usdc,
+        ITokenMessenger _cctpTokenMessenger,
+        address _l2Gateway,
+        IOvm_SpokePool _spokePool
+    ) WithdrawalAdapter(_l2Usdc, _cctpTokenMessenger, _l2Gateway) {
```

Comment:
https://github.com/across-protocol/contracts/pull/605/files#r1773159772

Reply:
Yes. It's copied from the Spoke Pool.